### PR TITLE
Implement simple clamping

### LINKED ISSUE #96: Run test_clamping 1 & 2 protocols
Part of #23.

- [x] Should be just a matter of implementing the `ClampVariable` action to delegate to `ModelEquation`, providing syntactic sugar.

Tested in `test_algebraic_models.py`. Which means we also need:
- [x] Handle models with no free variable in code generation.

- [x] The `var = var` equation that clamping to initial value results in gets simplified to just `BooleanTrue` by sympy. We might need to represent this differently in sympy, and/or handle it specially in `_add_or_replace_equations`. I suspect the best option is only to treat clamping to fixed value as syntactic sugar, and keep a separate action for clamping to initial value. All such actions can then be processed separately just before or after `_add_or_replace_equations`.
    - Question: if the RHS of the existing equation defining `var` refers to other variables, should this be allowed?

Possible extension:
- [ ] Check that a variable isn't defined more than once in the interface. Have a check before making changes that we don't have 2 define and/or clamp clauses targeting the same variable, and produce a friendly error.

### PR DESCRIPTION
**Pull request by jonc125**
Fixes #96, at least the core functionality needed for the clamping test protocols.

### CODE REVIEW

**Comment by jonc125** [comment]
I'm not sure why Travis isn't triggering for PRs that target another branch :(

**Comment by jonc125** [comment]
> I'm not sure why Travis isn't triggering for PRs that target another branch :(

Figured it out!

File: .travis.yml [hunk 1]
```diff
@@ -6,10 +6,6 @@ addons:
   apt:
     packages:
       - libsundials-serial-dev
-branches:
```
Comment (MichaelClerx) [inline .travis.yml]: I think this was here to prevent PRs from being built twice (once as branch, once as PR)

File: fc/parsing/actions.py [hunk 2]
```diff
@@ -900,7 +961,21 @@ def _add_or_replace_equations(self):
...
-        # TODO: Check units of newly added equations; apply conversions where needed?
+        # TODO: Check units of newly added equations; apply conversions where needed now or later?
+
+    def _handle_clamping(self):
+        """Clamp requested variables to their initial values."""
+        for clamp in self._clamps:
+            var = self.model.get_symbol_by_ontology_term(clamp.rdf_term)
+            defn = self.model.get_definition(var)
+            if var.initial_value is None:
+                value = defn.rhs.evalf()  # TODO: What if there are variable references in the RHS?
```
Comment (MichaelClerx) [inline fc/parsing/actions.py]: Is this for a future test case? Will evalf() fail here at least?

File: fc/parsing/actions.py [hunk 3]
```diff
@@ -956,6 +1034,9 @@ def _purge_unused_mathematics(self, output_symbols):
         # Remove them and their definitions
         for symbol in unused_symbols:
             self.model.remove_variable(symbol)
+        # Add time back in to the graph if needed
+        if time_symbol not in self.model.graph.nodes:
```
Comment (MichaelClerx) [inline fc/parsing/actions.py]: Why would it have disappeared?

Comment (jonc125) [reply fc/parsing/actions.py]: If the model modifications make the model algebraic (i.e. no ODEs) - which is the case for the clamping protocols.

File: fc/parsing/actions.py (hunk 2 above)
Comment (jonc125) [reply fc/parsing/actions.py]: Yes, I think evalf would fail in this case. It was the question I posed on #96.

File: .travis.yml (hunk 1 above)
Comment (jonc125) [reply .travis.yml]: Yeah, unfortunately it has an undesirable side effect! We can turn off branch builds in the Travis web config, but then you don't end up with builds of master for readme badges etc I think. Or we could turn off normal branch builds and have a cron build of master each day?

File: fc/parsing/actions.py (hunk 3 above)
Comment (MichaelClerx) [reply fc/parsing/actions.py]: Interesting! How does that happen? The remove_variable call above shouldn't ever remove the time variable, and I don't see any remove_variable calls in cellmlmanip itself?

Comment (jonc125) [reply fc/parsing/actions.py]: It's not that the variable is removed, just that it doesn't appear in model.graph as it isn't referenced in any equations.

File: .travis.yml (hunk 1 above)
Comment (MichaelClerx) [reply .travis.yml]: Hmmm... The `only: -master` is meant to make master build anyway. We use the same config without problems in PINTS

Comment (jonc125) [reply .travis.yml]: Yeah, having that config worked for that. But what it broke was PRs only got built if they were against master, not if they were building on top of another branch.

File: fc/parsing/actions.py (hunk 3 above)
Comment (MichaelClerx) [reply fc/parsing/actions.py]: Oh right, cause the graph gets rebuilt after the manipulations. Wonder if we should fix it at cellmlmanip level that the time variable is in there?

File: .travis.yml (hunk 1 above)
Comment (MichaelClerx) [reply .travis.yml]: Oh yeah. I've seen that behaviour too!

File: fc/parsing/actions.py (hunk 3 above)
Comment (jonc125) [reply fc/parsing/actions.py]: I wondered about that. It is of course possible for a CellML file not to have time at all! And if it didn't have any ODEs in the maths, I'm not sure how you'd tell what to consider time to be?

Comment (jonc125) [reply fc/parsing/actions.py]: Whereas here I know, because it was either in the original model or the protocol created it.

**Pull request merged**